src/systems/spacecraft_thruster_model/SpacecraftThrusterModel.cc

```diff
@@ -0,0 +1,431 @@
+  }
+  else
+  {
+    gzerr << "Please specify actuator "
+          << this->dataPtr->actuatorNumber <<" maxThrust.\n";
+  }
+
+  if (sdfClone->HasElement("dutyCycleFrequency"))
+  {
+    this->dataPtr->dutyCycleFrequency =
+          sdfClone->GetElement("dutyCycleFrequency")->Get<double>();
+  }
+  else
+  {
+    gzerr << "Please specify actuator "
+          << this->dataPtr->actuatorNumber <<" dutyCycleFrequency.\n";
+  }
+
+  sdfClone->Get<std::string>("commandSubTopic",
+      this->dataPtr->commandSubTopic, this->dataPtr->commandSubTopic);
+
+  // Subscribe to actuator command messages
+  std::string topic = transport::TopicUtils::AsValidTopic(
+      this->dataPtr->robotNamespace + "/" + this->dataPtr->commandSubTopic);
+  if (topic.empty())
+  {
+    gzerr << "Failed to create topic for [" << this->dataPtr->robotNamespace
+           << "]" << std::endl;
+    return;
+  }
+  else
+  {
+    gzdbg << "Listening to topic: " << topic << std::endl;
+  }
+  this->dataPtr->node.Subscribe(topic,
+      &SpacecraftThrusterModelPrivate::OnActuatorMsg, this->dataPtr.get());
+}
+
+//////////////////////////////////////////////////
+void SpacecraftThrusterModelPrivate::OnActuatorMsg(
+    const msgs::Actuators &_msg)
+{
+  std::lock_guard<std::mutex> lock(this->recvdActuatorsMsgMutex);
+  this->recvdActuatorsMsg = _msg;
+}
+
+//////////////////////////////////////////////////
+void SpacecraftThrusterModel::PreUpdate(const UpdateInfo &_info,
+    EntityComponentManager &_ecm)
+{
+  GZ_PROFILE("SpacecraftThrusterModel::PreUpdate");
+  // \TODO(anyone) Support rewind
+  if (_info.dt < std::chrono::steady_clock::duration::zero())
+  {
+    gzwarn << "Detected jump back in time ["
+        << std::chrono::duration_cast<std::chrono::seconds>(_info.dt).count()
+        << "s]. System may not work properly." << std::endl;
+  }
+
+  // If the joint or links haven't been identified yet, look for them
+  if (this->dataPtr->jointEntity == kNullEntity)
+  {
+    this->dataPtr->jointEntity =
+        this->dataPtr->model.JointByName(_ecm, this->dataPtr->jointName);
+
+    const auto parentLinkName = _ecm.Component<components::ParentLinkName>(
+        this->dataPtr->jointEntity);
+    this->dataPtr->parentLinkName = parentLinkName->Data();
+  }
+
+  if (this->dataPtr->linkEntity == kNullEntity)
+  {
+    this->dataPtr->linkEntity =
+        this->dataPtr->model.LinkByName(_ecm, this->dataPtr->linkName);
+  }
+
+  if (this->dataPtr->parentLinkEntity == kNullEntity)
+  {
+    this->dataPtr->parentLinkEntity =
+        this->dataPtr->model.LinkByName(_ecm, this->dataPtr->parentLinkName);
+  }
+
+  if (this->dataPtr->jointEntity == kNullEntity ||
+      this->dataPtr->linkEntity == kNullEntity ||
+      this->dataPtr->parentLinkEntity == kNullEntity)
+    return;
+
+  // skip UpdateForcesAndMoments if needed components are missing
+  bool doUpdateForcesAndMoments = true;
+
+  if (!_ecm.Component<components::WorldPose>(this->dataPtr->linkEntity))
```

Comment:
I think you can do these checks in the Configure stage as well.

Reply:
I moved the suggested checks to Configure, expecting that these components are not updated on runtime.